053-configurable-event-store/proposal.md

```diff
@@ -0,0 +1,457 @@
+   `ci.concourse-ci.org`, build events comprise ~70% of the total ~13GB used.
+   While storage is cheap, I can imagine this could result in general sluggishness
+   in larger Concourse deployments as indexes get large. While much of this can be
+   alleviated by more aggressive build log reaping, that can introduce performance
+   problems of its own, since deleting lots of build events can be taxing on
+   Postgres. I'm sure it's fairly obvious at this point that I really don't know
+   much about running Postgres/Concourse at scale, so let me know if what I'm
+   saying here doesn't hold true.
+1. Postgres wasn't designed for handling huge amounts of text data. Again, I
+   don't have much evidence/expertise to back this up, but having many large
+   rows in a `build_events` table strikes me as sub-optimal (compared to a system
+   designed for handling large amounts of text).
+1. Storing build events in a system designed for handling text data (like
+   Elasticsearch, for instance) allows you to do some pretty cool querying that
+   isn't feasible in Postgres.
+1. Gaining observability into build events isn't great. The syslog drainer
+   functionality mostly addresses this usecase, but isn't ideal IMO - it runs
+   on an interval (rather than continuously), and if you syslog forward build
+   events to an external store, you can choose to either store the build events in
+   two places, or reap the build events from Postgres (meaning you won't be able
+   to view the logs after).
+
+Some discussion around this motivation can be found in this discussion: concourse/concourse#5306.
+
+The second motivation is to introduce a general mechanism for processing build
+events. I can foresee having a configurable data transformation pipeline for
+build events that modifies events before storing them. I have a few ideas about
+some event processing that could be useful:
+
+1. Secret redaction - while we already have this feature, I think it's
+   implementation could be simplified by thinking of it as a data
+   transformation step (i.e. it receives log events and emits new log events with
+   secrets redacted), rather than trying to redact secrets on the fly.
+1. Server-side ANSI processing, à la concourse/concourse#3879
+
+Additionally, we could introduce behaviour like continuous syslog forwarding
+and migrating completed builds to cold storage fairly easily. This is explored
+more in [`EventStore` Composition](#event-store-composition).
+
+
+# Proposal
+
+To support having custom event stores, introduce an `EventStore` interface:
+
+```go
+type Key interface{}
+
+type EventStore interface {
+    Setup() error
+
+    Initialize(build db.Build) error
+    Finalize(build db.Build) error
+
+    Put(build db.Build, events []atc.Event) error
+    Get(build db.Build, requested int, cursor *Key) ([]event.Envelope, error)
+
+    Delete(builds []db.Build) error
+    DeletePipeline(pipeline db.Pipeline) error
```

Comment:
it got me wondering if we should really have these :thinking: from what I understand, the distinction between `Delete` and `Delete(Pipeline|Team)` is simply that the the one implementing the interface could do so in a more performant manner. Buut .. the Pipeline and Team concepts seem to be *too* Concourse specific, kinda "leaking" into the interface here.

tbh, I don't have a suggestion :sweat_smile: just seemed like overloading the interface a bit

Reply:
Yeah I agree, but couldn't really think of a better way to do it without losing the ability to `DROP` tables...

We *could* try to be smart with the Postgres implementation of the `EventStore`, where if we call `Delete(builds)` where `builds` exactly matches the list of pipeline builds, drop the table instead - though I feel like this could be finicky.